### PR TITLE
[WIP] Added GetUsage for filter by account ID, principal ID, start date and limit

### PR DESCRIPTION
**Pull request by AmudaPalani**
## Proposed changes

Added GetUsage for filter by account ID, principal ID, start date and limit

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (changes to code, which do not change application behavior)

## Checklist

- [x] I have filled out this PR template
- [x] I have read the [CONTRIBUTING](https://github.com/Optum/dce/blob/master/docs/CONTRIBUTING.md) doc
- [x] I have added automated tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (`README.md`, inline comments, etc.)
- [x] I have updated the `CHANGELOG.md` under a `## next` release, with a short summary of my changes

### CODE REVIEW

File: cmd/lambda/usage/list.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+		nextURL := buildNextURL(r, result.NextKeys)
+		w.Header().Add("Link", fmt.Sprintf("<%s>; rel=\"next\"", nextURL))
+	}
+
+	json.NewEncoder(w).Encode(usageResponseItems)
+}
+
+// parseGetUsageInput creates a GetUsageInput from the query parameters
+func parseGetUsageInput(r *http.Request) (usage.GetUsageInput, error) {
+	query := usage.GetUsageInput{
```
Comment (nathanagood) [inline cmd/lambda/usage/list.go]: We have a number of these "parse..Input()" methods. I'm wondering if we can make these a bit more dynamic and use reflection? It looks like this is already part of the gorilla toolkit (see https://stackoverflow.com/questions/40564842/convert-url-query-map-of-slices-to-struct-golang).

File: cmd/lambda/usage/list.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	nextPrincipalID := r.FormValue(NextPrincipalIDParam)
+	if len(nextPrincipalID) > 0 {
+		query.StartKeys["PrincipalId"] = nextPrincipalID
+	}
+
+	return query, nil
+}
+
+// buildNextURL merges the next parameters into the request parameters and returns an API URL.
+func buildNextURL(r *http.Request, nextParams map[string]string) string {
```
[the remark this answers is not in the document]
Comment (kddejong) [reply cmd/lambda/usage/list.go]: Very particular to the API layer so I say put it into the api response helpers package.

File: cmd/lambda/usage/list.go (hunk 1 above)
Comment (kddejong) [reply cmd/lambda/usage/list.go]: I'm going to be submitting a pull request very shortly that does this.

Comment (AmudaPalani) [reply cmd/lambda/usage/list.go]: Thanks

Comment (kddejong) [reply cmd/lambda/usage/list.go]: See https://github.com/kddejong/dce/blob/Feature/AccountControllerRedux/pkg/model/account.go#L4-L13
and
https://github.com/kddejong/dce/blob/Feature/AccountControllerRedux/cmd/lambda/accounts/list.go#L45
then
https://github.com/kddejong/dce/blob/8c9ff115c26ebe3938d1fd94902da0e5d88b3e1a/pkg/api/helpers.go#L12

File: cmd/lambda/usage/list.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (AmudaPalani) [reply cmd/lambda/usage/list.go]: Fixed

**Comment by AmudaPalani** [comment]
run dce#destory

**Pull request closed without merging**